src/cmd/run.js

```diff
@@ -186,87 +187,70 @@ export default function run(
-                  'add-on was already installed');
+  let manifestData = await getValidatedManifest(sourceDir);
+
+  runner = new ExtensionRunner({
+    sourceDir,
+    firefox,
+    firefoxBinary,
+    manifestData,
+    profilePath: firefoxProfile,
+  });
+
+  profile = await  runner.getProfile();
+
+  if (!preInstall) {
+    log.debug('Deferring extension installation until after ' +
+              'connecting to the remote debugger');
+  } else {
+    log.debug('Pre-installing extension as a proxy file');
+    addonId = await runner.installAsProxy(profile);
+    installed = true;
+  }
+
+  runningFirefox = await runner.run(profile);
+
+  if (installed) {
+    log.debug('Not installing as temporary add-on because the ' +
+              'add-on was already installed');
+  } else if (requiresRemote) {
+    client = await firefoxClient();
+
+    try {
+      addonId = await runner.installAsTemporaryAddon(client).then(
+        (installResult: FirefoxRDPResponseAddon) => installResult.addon.id
+      );
+    } catch (error) {
+      if (error instanceof RemoteTempInstallNotSupported) {
+        log.debug(`Caught: ${error}`);
+        throw new WebExtError(
+          'Temporary add-on installation is not supported in this version ' +
+          'of Firefox (you need Firefox 49 or higher). For older Firefox ' +
+          'versions, use --pre-install');
       } else {
-        return runner.installAsTemporaryAddon(client)
-          .then((installResult: FirefoxRDPResponseAddon) => {
-            addonId = installResult.addon.id;
-          });
+        throw error;
       }
-    })
-    .catch(onlyInstancesOf(RemoteTempInstallNotSupported, (error) => {
-      log.debug(`Caught: ${error}`);
-      throw new WebExtError(
-        'Temporary add-on installation is not supported in this version ' +
-        'of Firefox (you need Firefox 49 or higher). For older Firefox ' +
-        'versions, use --pre-install');
-    }))
-    .then(() => {
-      if (noReload) {
-        log.debug('Extension auto-reloading has been disabled');
-      } else {
-        if (!addonId) {
-          throw new WebExtError(
-            'Unexpected missing addonId in the installAsTemporaryAddon result'
-          );
-        }
-
-        log.debug(
-          `Reloading extension when the source changes; id=${addonId}`);
-        reloadStrategy({
-          firefox: runningFirefox,
-          profile, client, sourceDir, artifactsDir, addonId,
-        });
+    }
+
+    if (noReload) {
+      log.debug('Extension auto-reloading has been disabled');
+    } else {
+      if (!addonId) {
+        throw new WebExtError(
+          'Unexpected missing addonId in the installAsTemporaryAddon result'
+        );
       }
-      return firefox;
-    });
+
+      log.debug(
+        `Reloading extension when the source changes; id=${addonId}`);
+      reloadStrategy({
+        firefox: runningFirefox,
+        profile, client, sourceDir, artifactsDir, addonId,
+      });
+    }
+  }
```

Comment:
agree, this function and the `establishConnection` helper above are my favorites "rewrited to async functions" use cases :-D

(once I had done with them, I re-looked at the diff and I couldn't block some "tears of happiness", like a young groupie at the concert of their favorite band :-P)